### PR TITLE
2203-2

### PR DESCRIPTION
**Pull request by imikailoby**
## Номер
2203

## Номер задания
2

## Ссылка на видео с демо
https://youtu.be/ZQVj3V2zVng

## Комментарии
1/3 - Done.
Пришлось отключить некоторые проверки reek/rubocop дабы паттерны прошли проверку

### CODE REVIEW

**Comment by anatoliliotych** [comment]
you have typo in commit name and I can see what was implemented using my own eyes.

File: 2203/2/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+text/*
```
Comment (anatoliliotych) [inline 2203/2/.gitignore]: unnecessary file.

File: 2203/2/README.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
+Regular setup
+--------------
+Для установки прекрасной утилиты для подсчета различной интересной статистики вам потребуются:
+
+* `rvm` либо `rbenv`
```
Comment (anatoliliotych) [inline 2203/2/README.md]: I'd prefer to use bundler with Gemfile.

File: 2203/2/fix_words.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    files_path = Dir['text/*'].sort.map { |filename| filename.partition('text/ ').last }
+    files_path.map { |filename| filename.partition(/( против | vs )/i).first }.uniq
+  end
+
+  def self.search_rapers_battle(raper_name)
+    Dir.entries('text/').select { |files| /\b#{raper_name} (против|vs)\b/i =~ files }
+  end
+
+  def self.hash_with_bad_words(line, battle_words_count, battle)
+    line.split(' '). select do |bad_word|
```
Comment (anatoliliotych) [inline 2203/2/fix_words.rb]: it's strange usage of select

File: 2203/2/fix_words.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+      battle_words_count[battle] += 1 if bad_word.include?('*') || RussianObscenity.obscene?(bad_word)
```
Comment (anatoliliotych) [inline 2203/2/fix_words.rb]: you can ignore such conditions by exiting from iteration earlier

File: 2203/2/fix_words.rb [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    end
+  end
+
+  def self.hash_with_all_words(line, battle_words_count)
+    line.split(' ').select { |round_line| battle_words_count[round_line] += 1 }
+  end
+
+  def self.delete_small_words(word_count, line)
+    line.split(' ').select { |each_word| each_word.length >= 4 }.select { |words| word_count[words] += 1 }
```
Comment (anatoliliotych) [inline 2203/2/fix_words.rb]: can you join two selects?

File: 2203/2/versus.rb [hunk 6]
```diff
@@ -0,0 +1,112 @@
+require 'optparse'
+require 'terminal-table'
+require 'russian_obscenity'
+require './fix_words'
+
+# This class can show some intresting stats about russian battle rapers
+class Versus
```
Comment (anatoliliotych) [inline 2203/2/versus.rb]: what is the sense of this class? it looks like wrapper for your non OOP solution.

**Review by anatoliliotych — CHANGES_REQUESTED** [review]

File: 2203/2/README.md (hunk 2 above)
Comment (imikailoby) [reply 2203/2/README.md]: fixed

File: 2203/2/top_words_counter.rb [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    else
+      show_search_error(raper_name)
+    end
+  end
+
+  def self.show_search_error(raper_name)
+    puts "Я не знаю МЦ #{raper_name}. Зато мне известны:"
+    BattleChecker.rapers_list.each { |raper_names| puts raper_names }
+  end
+end
```
[the remark this answers is not in the document]
Comment (imikailoby) [reply 2203/2/top_words_counter.rb]: Hey, @houndci-bot! Already fixed.

**Comment by imikailoby** [comment]
@anatoliliotych recheck please

**Comment by imikailoby** [comment]
@anatoliliotych Can you recheck, please? I hope all it's fine. But I have some question about best way to minimize dir/file read/open operation in Ruby. I think I'm doesn't right in current code realization with this question (and with another code ha-ha). But I'm very thired and can't search and read more information about that. I'll hope you help me.

File: 2203/2/rapper_parser.rb [hunk 8]
```diff
@@ -0,0 +1,17 @@
+require './file_parser.rb'
+
+# This class parse all rappers
+class RapperParser
+  def self.list
+    rapper_list.sort.uniq
+  end
+
+  def self.rapper_list
```
Comment (anatoliliotych) [inline 2203/2/rapper_parser.rb]: should be private

File: 2203/2/versus.rb [hunk 9]
```diff
@@ -0,0 +1,22 @@
+require 'optparse'
+require 'terminal-table'
+require './russian_declensions.rb'
+require './bad_words_counter.rb'
+
+OptionParser.new do |options|
+  options.on('--top-bad-words=') do |number|
+    table = Terminal::Table.new do |row|
+      BadWordsCounter.sorting_bad_words[0...number.to_i].to_h.each do |name, bad_words_count|
```
Comment (anatoliliotych) [inline 2203/2/versus.rb]: you are still calling TopWordsStatistic for each rapper twice :)

**Review by anatoliliotych — CHANGES_REQUESTED** [review]

File: 2203/2/versus.rb (hunk 9 above)
Comment (imikailoby) [reply 2203/2/versus.rb]: @anatoliliotych fixed

File: 2203/2/rapper_parser.rb [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+class RapperParser
+  attr_reader :list
+
+  def initialize
+    @list = rapper_list.sort.uniq
+  end
+
+  def self.rapper_list
+    array = []
+    FileParser.file_list.each do |filename|
```
Comment (allomov) [inline 2203/2/rapper_parser.rb]: MAP!

**Comment by allomov** [comment]
parsers driven development

File: 2203/2/author_reader.rb [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+require './reader'
+
+# This class read rappers name
+class AuthorReader < Reader
+  private
+
+  def find_content(content)
+    result << Author.new(content[:name])
+    battles = content[:text].map { |text| Battle.new(text) }
+    result.last.add_battles(battles)
```
Comment (allomov) [inline 2203/2/author_reader.rb]: почему этот метод называется `find_content` ?

понятнее было бы написать такой код
```
new_author = Author.new(content[:name])
battles = content[:text].map { |text| Battle.new(text) }
new_author.add_battles(battles)
result << new_author
```
или еще вариант

```ruby
def find_contnet_или_как_этот_метод_называется(content)
  author = AuthorRegistry.find_or_create_by_name(content[:name])
  author.add_battles(battles_from_content(content[:text]))
end
def battles_from_contnet(text_arrays)
  battles = text_arrays.map { |text| Battle.new(text) }
end
```
что у тебя в `content[:text]`? как мне понять, что это массив?

File: 2203/2/battle.rb [hunk 12]
```diff
@@ -0,0 +1,24 @@
+require 'russian_obscenity'
+
+# This class show battles with stats
+class Battle
+  def initialize(battle_text)
+    @battle_text = battle_text
+  end
+
+  def all_words
+    @battle_text.scan(/[а-яА-Яa-zA-Z]+/)
```
Comment (allomov) [inline 2203/2/battle.rb]: константу мне! константу!

**Pull request closed without merging**